Sema: Fix existential type erasure around parameterized protocols

AnthonyLatsis opened the pull request:
Fixes rdar://132064309.

AnthonyLatsis committed:
Sema: Always use archetype-based type erasure for existential opening

AnthonyLatsis committed:
Sema: Fix handling of parameterized protocols in type erasure for existentials

AnthonyLatsis committed:
[NFC] AST: Consolidate some repeated `DependentMemberType` logic in `Type` methods

AnthonyLatsis committed:
[NFC] Sema: Remove unused generic parameter result from `canOpenExistentialCallArgument`

AnthonyLatsis committed:
Sema: Handle the edge case of in-out metatype parameter in `canOpenExistentialCallArgument`

AnthonyLatsis committed:
test: Actually check if we opened an existential call argument

AnthonyLatsis committed:
[NFC] AST: Clean up unused field and parameter in `LocalArchetypeType` subclasses

AnthonyLatsis committed:
[NFC] AST: Allow calling `OpenedArchetypeType::getAny` with an existing archetype

AnthonyLatsis commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented on lib/Sema/OpenedExistentials.cpp:
```diff
@@ -644,14 +637,11 @@ swift::canOpenExistentialCallArgument(ValueDecl *callee, unsigned paramIdx,
-
-  ASSERT(existentialTy->isAnyExistentialType());
-
+  // The binding could be an existential metatype. Get the instance type for
+  // conformance checks and to build an opened existential signature. If the
+  // instance type is not an existential type, i.e., the metatype is nested,
+  // bail out.
+  const Type existentialTy = bindingTy->getMetatypeInstanceType();
```
Looking through a metatype here causes us to open the existential in case 1, but not case 2:

```swift
func foo<T : BitwiseCopyable>(_: T) -> T {}

// 1
do {
  let exist: any Any.Type
  let result = foo(exist)
}

// 2
do {
  let exist: any Any.Type.Type
  let result = foo(exist)
}
```

I took care to not change the behaviour, but it clearly contradicts [SE-0352](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0352-implicit-open-existentials.md#avoid-opening-when-the-existential-type-satisfies-requirements-in-swift-5). I suppose this became a problem when metatypes started conforming to these marker protocols. @DougGregor Do you think we could risk correcting this?

Btw, the standard library [seems to be making use](https://github.com/swiftlang/swift/blob/f1f10517d5b6f4e07b316f53fb50f0f8b2928b6a/stdlib/public/core/KeyPath.swift#L3769) of the current case 1 behaviour.

AnthonyLatsis commented:
@swift-ci please test source compatibility release

slavapestov commented on lib/AST/ParameterPack.cpp:
```diff
@@ -176,17 +176,12 @@ void TypeBase::getTypeParameterPacks(
-
-  return t->isRootParameterPack();
+bool TypeBase::isParameterPack() const {
+  return getDependentMemberRoot()->isRootParameterPack();
```
s/getDependentMemberRoot()/getRootGenericParam()/

AnthonyLatsis replied:
`getRootGenericParam()` might return a null pointer. I don’t think a client would expect this `TypeBase` method to crash if not given something rooted on a generic parameter type.

AnthonyLatsis committed:
[NFC] OpenedArchetypeType: Rename `getAny` to `openAnyExistentialType`

AnthonyLatsis committed:
[NFC] ConstraintSystem: Rename `openExistentialType` to `openAnyExistentialType`

This method can open an existential metatype too, and we already use
`isAnyExistentialType` to mean "existential type or existential
metatype"

AnthonyLatsis committed:
AST: Fix the superclasses of some type nodes

AnthonyLatsis committed:
TypeCheckType: Do not form protocol type with generic parent in `resolveTypeInContext`

AnthonyLatsis commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented:
@swift-ci please test source compatibility debug
